test/quantization/core/test_quantized_tensor.py

```diff
@@ -328,6 +329,82 @@ def test_qtensor_dtypes(self):
+            scales = torch.rand(2).abs()
+            zero_points = (torch.rand(2) * 10).round().to(zero_type)
+
+            # dqr = torch.quantize_per_channel(r, scales, zero_points, axis, dtype)
+            # dqr = dqr.to(device)
+            # dqr_cuda = torch.quantize_per_channel(r.to(device), scales.to(
+            #     device), zero_points.to(device), axis, dtype)
+            # dqr_cuda = dqr_cuda.to('cpu')
+            # self.assertEqual('cuda', dqr.device.type)
+            # self.assertEqual('cpu', dqr_cuda.device.type)
+
+            if zero_type == torch.long:  # per_tensor doesn't have float support for zero_point
+                scale = scales.max().item()
+                zero_point = zero_points.max().item()
+                qr = torch.quantize_per_tensor(r, scale, zero_point, dtype)
+                qr = qr.to(device)
+                qr_cuda = torch.quantize_per_tensor(r.to(device), scale, zero_point, dtype)
+                qr_cuda = qr_cuda.to('cpu')
+                self.assertEqual('cuda', qr.device.type)
+                self.assertEqual('cpu', qr_cuda.device.type)
+
+    @unittest.skipIf(not torch.cuda.is_available() or TEST_WITH_ROCM, 'CUDA is not available')
+    def test_compare_quant_dequant_device_numerics(self):
+        dtype_and_zero_types = [
+            (torch.quint8, torch.float),
+            (torch.qint8, torch.float),
+            #  (torch.qint32, torch.float) not supported for quantize_per_channel
+            (torch.quint8, torch.long),
+            (torch.qint8, torch.long),
+            (torch.qint32, torch.long),
+        ]
+        axis = 1
+        device = torch.device('cuda')
+        for i in range(20):
+            for dtype, zero_type in dtype_and_zero_types:
+                r = torch.rand(2, 2) * 10
+                r[0, 0] = 2.5
+                scales = torch.rand(2).abs()
+                zero_points = (torch.rand(2) * 10).round().to(zero_type)
+
+                # qr = torch.quantize_per_channel(r, scales, zero_points, axis, dtype)
+                # dqr = qr.dequantize()
+                # qr_cuda = torch.quantize_per_channel(r.to(device), scales.to(
+                #     device), zero_points.to(device), axis, dtype)
+                # dqr_cuda = qr_cuda.dequantize()
+                # self.assertEqual(qr, qr_cuda)
+                # self.assertEqual(qr.int_repr(), qr_cuda.int_repr())
+                # self.assertEqual(dqr, dqr_cuda)
+
+                if zero_type == torch.long:  # per_tensor doesn't have support for Float zero_point
+                    scale = scales.max().item()
+                    zero_point = zero_points.max().item()
+                    qtr = torch.quantize_per_tensor(r, scale, zero_point, dtype)
+                    dqtr = qtr.dequantize()
+                    qtr_cuda = torch.quantize_per_tensor(r.to(device), scale, zero_point, dtype)
+                    dqtr_cuda = qtr_cuda.dequantize()
+                    self.assertEqual(qtr, qtr_cuda)
```

Comment:
Since you are trying with random data on different cores, I wonder if assertEqual will always work. Do you need a check for closeness here?, i.e something like
```
     self.assertTrue(np.allclose(qr.int_repr(), ref_res))
     self.assertTrue(np.allclose(r.numpy(), rqr.numpy(), atol=2 / np.min(scales.numpy())))
```

Reply:
I can make that change but as it stands the implementations are exactly equal on cuda and cpu.